kong/db/declarative/init.lua

```diff
@@ -0,0 +1,295 @@
+                " supported): " .. filename
+  end
+
+  if not dc_table then
+    return nil, "failed parsing declarative configuration file " ..
+        filename .. (err and ": " .. err or "")
+  end
+
+  local ok, err_t = self.schema:validate(dc_table)
+  if not ok then
+    return nil, pretty_print_error(err_t), err_t
+  end
+
+  local entities
+  entities, err_t = self.schema:flatten(dc_table)
+  if err_t then
+    return nil, pretty_print_error(err_t), err_t
+  end
+
+  return entities
+end
+
+
+function declarative.to_yaml_string(entities)
+  local pok, yaml, err = pcall(lyaml.dump, {entities})
+  if not pok then
+    return nil, yaml
+  end
+  if not yaml then
+    return nil, err
+  end
+
+  -- drop the multi-document "---\n" header and "\n..." trailer
+  return yaml:sub(5, -5)
+end
+
+
+function declarative.to_yaml_file(entities, filename)
+  local yaml, err = declarative.to_yaml_string(entities)
+  if not yaml then
+    return nil, err
+  end
+
+  local fd, err = io.open(filename, "w")
+  if not fd then
+    return nil, err
+  end
+
+  local ok, err = fd:write(yaml)
+  if not ok then
+    return nil, err
+  end
+
+  fd:close()
+
+  return true
+end
+
+
+function declarative.load_into_db(dc_table)
+  assert(type(dc_table) == "table")
+
+  local schemas = {}
+  for entity_name, _ in pairs(dc_table) do
+    table.insert(schemas, kong.db[entity_name].schema)
+  end
+  local sorted_schemas, err = topological_sort(schemas)
+  if not sorted_schemas then
+    return nil, err
+  end
+
+  local schema, primary_key, ok, err, err_t
+  for i = 1, #sorted_schemas do
+    schema = sorted_schemas[i]
+    for _, entity in pairs(dc_table[schema.name]) do
+      entity = deepcopy(entity)
+      entity._tags = nil
+
+      primary_key = schema:extract_pk_values(entity)
+
+      ok, err, err_t = kong.db[schema.name]:upsert(primary_key, entity)
+      if not ok then
+        return nil, err, err_t
+      end
+    end
+  end
+
+  return true
+end
+
+
+local function remove_nulls(tbl)
+  for k,v in pairs(tbl) do
+    if v == null then
+      tbl[k] = nil
+    elseif type(v) == "table" then
+      tbl[k] = remove_nulls(v)
+    end
+  end
+  return tbl
+end
+
+
+function declarative.load_into_cache(entities)
+
+  -- FIXME atomicity of cache update
+  kong.cache:purge()
+
+  for entity_name, items in pairs(entities) do
+    local dao = kong.db[entity_name]
+    local schema = dao.schema
+
+    local uniques = {}
+    local page_for = {}
+    local foreign_fields = {}
+    for fname, fdata in schema:each_field() do
+      if fdata.unique then
+        table.insert(uniques, fname)
+      end
+      if fdata.type == "foreign" then
+        page_for[fdata.reference] = {}
+        foreign_fields[fname] = fdata.reference
+      end
+    end
+
+    local ids = {}
+    for id, item in pairs(items) do
+      table.insert(ids, id)
+
+      local cache_key = dao:cache_key(id)
+      item = remove_nulls(item)
+      kong.cache:get(cache_key, nil, function()
```

Comment:
No error handling here? Also, we should populate from the mlcache instance directly, which seems more correct (and will also avoid error messages such as `"failed to get from node cache: ..."` in cases of failures during db-less writes.